### ISSUE TITLE
[python] unit test `test_query_interrupt.py` sending `KeyboardInterrupt` to wrong target on aarch64-darwin sometimes

### ISSUE THREAD

**Issue by ghost** [duckdb/duckdb#10702]
### What happens?

TL;DR `tests/fast/api/test_query_interrupt.py` is racy and can send keyboard interrupts to random other tests.

when running the python unit tests on aarch64-darwin 3 / 5 times they fail due to a `KeyboardInterrupt`. https://github.com/NixOS/nixpkgs/pull/288913#issuecomment-1947513507

logs from the CI builds:
 [fail 1 build log](https://logs.ofborg.org/?key=nixos/nixpkgs.288913&attempt_id=ccabe88f-440b-42c3-b896-645454eb2173) [fail 2 build log](https://logs.ofborg.org/?attempt_id=8f7190f3-c49d-48d1-ab45-811448ac38a3&key=nixos%2Fnixpkgs.288913) [fail 3 build log](https://logs.ofborg.org/?key=nixos/nixpkgs.288913&attempt_id=fb319bd8-c295-4033-8c38-8dbff3140514)

Note: these failures just appear on aarch64-darwin, the other CI builds on x64-linux, x64-darwin and aarch64-linux have not yet failed.

```console
tests/fast/api/test_read_csv.py::TestReadCSV::test_dtype_as_list
  /nix/store/14a99rbrb350d4zp49ds09isyssm5ksm-python3.11-pytest-7.4.4/lib/python3.11/site-packages/_pytest/runner.py:262: PluggyTeardownRaisedWarning: A plugin raised an exception during an old-style hookwrapper teardown.
  Plugin: capturemanager, Hook: pytest_runtest_setup
  KeyboardInterrupt: 
  For more information see https://pluggy.readthedocs.io/en/stable/api_reference.html#pluggy.PluggyTeardownRaisedWarning
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
<frozen codecs>:319: KeyboardInterrupt
(to show a full traceback on KeyboardInterrupt use --full-trace)
```

```console
__________________________ TestReadCSV.test_skiprows ___________________________
KeyboardInterrupt
The above exception was the direct cause of the following exception:
self = <test_read_csv.TestReadCSV object at 0x11c044ad0>
duckdb_cursor = <duckdb.duckdb.DuckDBPyConnection object at 0x12b14d170>
    def test_skiprows(self, duckdb_cursor):
        rel = duckdb_cursor.read_csv(TestFile('category.csv'), skiprows=1)
>       res = rel.fetchone()
E       RuntimeError: Query interrupted
tests/fast/api/test_read_csv.py:106: RuntimeError
```

```console
tests/fast/api/test_streaming_result.py ..s.                             [ 43%]
tests/fast/api/test_to_csv.py ...........
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
/nix/store/flkpbg20pfzfa9f2lwzrzjl718i2ccw7-python3-3.11.7/lib/python3.11/tempfile.py:368: KeyboardInterrupt
(to show a full traceback on KeyboardInterrupt use --full-trace)
```

### To Reproduce

the unit tests are built using the  tag 0.10.0 with one patch https://github.com/duckdb/duckdb/commit/3d7aa3ed46ecf5f18122559e385b75f1f5e9aba8.patch, BUILD_HTTPFS = 1 and extension autoinstall / autoloading disabled by default for the nixpkgs duckdb build. https://github.com/NixOS/nixpkgs/pull/288913

running the python unit tests fail 60% of the time (n=5) with a KeyboardInterrupt on aarch64-darwin.

note -- the python unit tests are run while the cmake c++ binary is building

### OS:

darwin aarch64 nixpkgs

### DuckDB Version:

0.10.0, 0.9.2

### DuckDB Client:

python

### Full Name:

annalee

### Affiliation:

none (packaging for nixpkgs / NixOS)

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have not tested with any build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by a-n-n-a-l-e-e** [comment]
setup a PR https://github.com/NixOS/nixpkgs/pull/289199 to test if it fails on 0.9.2 on the CI and it does [fail](https://logs.ofborg.org/?key=nixos/nixpkgs.289199&attempt_id=6a97c3fc-a9f4-4b67-8eed-55f034d395a7) on 1/1 try.

```console
tests/fast/api/test_query_interrupt.py .                                 [ 49%]
[429/615] Building CXX object extension/tpcds/CMakeFiles/tpcds_loadable_extension.dir/tpcds_extension.cpp.o
tests/fast/api/test_read_csv.py .......
=============================== warnings summary ===============================
tests/fast/api/test_read_csv.py::TestReadCSV::test_delimiter
  /nix/store/14a99rbrb350d4zp49ds09isyssm5ksm-python3.11-pytest-7.4.4/lib/python3.11/site-packages/_pytest/runner.py:262: PluggyTeardownRaisedWarning: A plugin raised an exception during an old-style hookwrapper teardown.
  Plugin: capturemanager, Hook: pytest_runtest_teardown
  KeyboardInterrupt: 
  For more information see https://pluggy.readthedocs.io/en/stable/api_reference.html#pluggy.PluggyTeardownRaisedWarning
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

note that the python unit tests are running in parallel to the normal duckdb build.

**Comment by a-n-n-a-l-e-e** [comment]
another, this one from 0.10.0 aarch64-darwin [log](https://logs.ofborg.org/?key=nixos/nixpkgs.288913&attempt_id=3eab32c8-b36e-40d3-ac20-24afcc228a7b)
```console
tests/fast/api/test_query_interrupt.py .                                 [ 41%]
tests/fast/api/test_read_csv.py .......
=============================== warnings summary ===============================
tests/fast/api/test_read_csv.py::TestReadCSV::test_delimiter_and_sep
  /nix/store/14a99rbrb350d4zp49ds09isyssm5ksm-python3.11-pytest-7.4.4/lib/python3.11/site-packages/_pytest/runner.py:262: PluggyTeardownRaisedWarning: A plugin raised an exception during an old-style hookwrapper teardown.
  Plugin: capturemanager, Hook: pytest_runtest_setup
  KeyboardInterrupt: 
  For more information see https://pluggy.readthedocs.io/en/stable/api_reference.html#pluggy.PluggyTeardownRaisedWarning
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
<frozen codecs>:319: KeyboardInterrupt
(to show a full traceback on KeyboardInterrupt use --full-trace)
```

**Comment by a-n-n-a-l-e-e** [comment]
figured out this is due to `test_query_interrupt` -- thought we disabled it but that was a different racy test, `test_connection_interrupt`. anyway:

```python
def send_keyboard_interrupt():
    # Wait a little, so we're sure the 'execute' has started
    time.sleep(0.1)
    # Send an interrupt to the main thread
    thread.interrupt_main()


class TestQueryInterruption(object):
    def test_query_interruption(self):
        con = duckdb.connect()
        thread = threading.Thread(target=send_keyboard_interrupt)
        # Start the thread
        thread.start()
        try:
            res = con.execute('select count(*) from range(1000000000)').fetchall() # <--- this can finish before sigint is sent.
        except RuntimeError:
            # If this is not reached, we could not cancel the query before it completed
            # indicating that the query interruption functionality is broken
            assert True
        except KeyboardInterrupt:
            pytest.fail()
```
